### DISCUSSION TITLE
In Section 4 Lesson 15, there is an error in the suggested code for detecting duplicate users during the repair process

### DISCUSSION THREAD

**Discussion by kildren-coder** [Cyfrin/security-and-auditing-full-course-s23#198] [Q&A]
At the end of the [video](https://updraft.cyfrin.io/courses/security/puppy-raffle/dos-mitigation), we need to provide suggestions for detecting duplicate users. The approach in the video is correct, but the code provided contains errors.

The [code](https://github.com/Cyfrin/4-puppy-raffle-audit/blob/audit-data/audit-data/2023-09-01-puppy-raffle.md#m-1-looping-through-players-array-to-check-for-duplicates-in-puppyraffleenterraffle-is-a-potential-dos-vector-incrementing-gas-costs-for-future-entrants) is as follows:
```diff
+    mapping(address => uint256) public addressToRaffleId;
+    uint256 public raffleId = 0;
    .
    .
    .
    function enterRaffle(address[] memory newPlayers) public payable {
        require(msg.value == entranceFee * newPlayers.length, "PuppyRaffle: Must send enough to enter raffle");
        for (uint256 i = 0; i < newPlayers.length; i++) {
            players.push(newPlayers[i]);
+            addressToRaffleId[newPlayers[i]] = raffleId;            
        }

-        // Check for duplicates
+       // Check for duplicates only from the new players
+       for (uint256 i = 0; i < newPlayers.length; i++) {
+          require(addressToRaffleId[newPlayers[i]] != raffleId, "PuppyRaffle: Duplicate player");
+       }    
-        for (uint256 i = 0; i < players.length; i++) {
-            for (uint256 j = i + 1; j < players.length; j++) {
-                require(players[i] != players[j], "PuppyRaffle: Duplicate player");
-            }
-        }
        emit RaffleEnter(newPlayers);
    }
.
.
.
    function selectWinner() external {
+       raffleId = raffleId + 1;
        require(block.timestamp >= raffleStartTime + raffleDuration, "PuppyRaffle: Raffle not over");
```

According to the given code, all players are checked after being added to the array and ```addressToRaffleId[newPlayers[i]] = raffleId``` is set. Then, checking ```require(addressToRaffleId[newPlayers[i]] != raffleId, "PuppyRaffle: Duplicate player")``` obviously leads to a situation where the require check cannot pass.

We expect that before a player is added to the queue, their raffleId should be checked to see if it matches the raffleId of the current round. If it matches, it means they have already joined this round of the raffle and should not be added again.

Considering the impact of default values, the correct code should be as follows:
```diff
+    mapping(address => uint256) public addressToRaffleId;
+    uint256 public raffleId = 1; // raffleId = 0 will lead to fail to add new player
    .
    .
    .
    function enterRaffle(address[] memory newPlayers) public payable {
        require(msg.value == entranceFee * newPlayers.length, "PuppyRaffle: Must send enough to enter raffle");
        for (uint256 i = 0; i < newPlayers.length; i++) {
+            require(addressToRaffleId[newPlayers[i]] != raffleId, "PuppyRaffle: Duplicate player");            
               players.push(newPlayers[i]);
+            addressToRaffleId[newPlayers[i]] = raffleId;            
        }

-        // Check for duplicates
-        for (uint256 i = 0; i < players.length; i++) {
-            for (uint256 j = i + 1; j < players.length; j++) {
-                require(players[i] != players[j], "PuppyRaffle: Duplicate player");
-            }
-        }
        emit RaffleEnter(newPlayers);
    }
.
.
.
    function selectWinner() external {
+       raffleId = raffleId + 1;
        require(block.timestamp >= raffleStartTime + raffleDuration, "PuppyRaffle: Raffle not over");
```

**Comment by EngrPips** [comment]
Hello @kildren-coder, This should be reported in the `issues` section of this Github and not here in `Discussion,` `Discussion` is primarily used to help people who get stuck. Please use the `issues` section to report this

**Reply by kildren-coder** [reply]
Thanks, my bad

**Comment by kildren-coder** [comment]
report in the [issue](https://github.com/Cyfrin/security-and-auditing-full-course-s23/issues/199)

**Reply by EngrPips** [reply]
Thanks.
